### PR TITLE
Flashing instructions

### PR DESCRIPTION
**Pull request by drewmoseley**
This PR was recently submitted to the Yocto mailing list and based on the post from @agherzan to suggest using Github PRs and Jenkins I figured I'd give that a shot here.

### CODE REVIEW

**Comment by agherzan** [comment]
I'm setting up the jenkins server as we speak. This will be the first PR testing it.

File: conf/layer.conf [hunk 1]
```diff
@@ -8,3 +8,6 @@ BBFILES += "${LAYERDIR}/recipes-*/*/*.bb \
 BBFILE_COLLECTIONS += "meta-chip"
 BBFILE_PATTERN_meta-chip = "^${LAYERDIR}/"
 BBFILE_PRIORITY_meta-chip = "6"
+
+# Add this to get access to sunxi-tools recipe
+LAYERDEPENDS_meta-chip = "meta-sunxi"
```
Comment (agherzan) [inline conf/layer.conf]: A new line would be nice here.

File: recipes-bsp/u-boot/u-boot-chip_git.bb [hunk 2]
```diff
@@ -19,7 +20,13 @@ SRC_URI = " \
...
+    install ${B}/spl/${SPL_ECC_BINARY} ${B}/${SPL_ECC_BINARY}
     install ${B}/spl/${SPL_BINARY} ${B}/${SPL_BINARY}
 }
 
 COMPATIBLE_MACHINE = "chip"
+
+do_deploy_append() {
+    install -m 644 ${B}/${SPL_ECC_BINARY} ${DEPLOYDIR}/${SPL_ECC_BINARY}-${PV}-${PR}
+    ln -sf ${SPL_ECC_BINARY}-${PV}-${PR} ${DEPLOYDIR}/${SPL_ECC_BINARY}
+}
```
Comment (agherzan) [inline recipes-bsp/u-boot/u-boot-chip_git.bb]: New line please.

**Comment by agherzan** [comment]
ok to test

File: README [hunk 3]
```diff
@@ -62,4 +62,93 @@ To build a machine supported by this BSP layer follow the next steps:
...
+UBI_IMAGE shell variable before invoking the generated script as
+follows:
+
+# UBI_IMAGE=chip-hwup-image-chip.ubi \
+        ./tmp/deploy/images/chip/flash_CHIP_board.sh
+
+This script will take some time to execute.  The output of this script
+as run on your host machine should look similar to the following:
+
+    # UBI_IMAGE=chip-hwup-image-chip.ubi ./build/chip/tmp/deploy/images/chip/flash_CHIP_board.sh
```
Comment (agherzan) [inline README]: A little bit of inconsistency here. We use a path which includes the build directory while above we use a path inside this directory. Would be nice to have them both with or without build directory.

File: README [hunk 4]
```diff
@@ -62,4 +62,93 @@ To build a machine supported by this BSP layer follow the next steps:
...
+As part of the build including this BSP layer, a U-Boot script and a
+shell script are created to assist in flashing your images to the CHIP
+board.  Note that this script assumes a maximum size for your UBI
+image of 167772160 (0x0A000000) bytes.  If your UBI image exceeds
+that, then you will need to adapt this to your environment.
+
+To successfully run this script you will first need to set your board
+into FEL mode. Insert a wire into the U14 header between pin 7 (FEL)
+and GND.  Then connect the board with a USB cable to your build
+system. More details here:
```
Comment (agherzan) [inline README]: I would reword this to something like "to your host".

**Comment by agherzan** [comment]
test this please

File: README [hunk 5]
```diff
@@ -62,4 +62,93 @@ To build a machine supported by this BSP layer follow the next steps:
 II. Flashing a C.H.I.P. board
 ========================================
 
-<TODO>
+As part of the build including this BSP layer, a U-Boot script and a
+shell script are created to assist in flashing your images to the CHIP
+board.  Note that this script assumes a maximum size for your UBI
```
Comment (agherzan) [inline README]: This text introduces from time to time double spaces. Would be nice to have them fixed.

File: recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb [hunk 6]
```diff
@@ -0,0 +1,89 @@
+SUMMARY = "U-boot boot scripts for CHIP boards"
+LICENSE = "MIT"
+LIC_FILES_CHKSUM = "file://${COREBASE}/meta/COPYING.MIT;md5=3da9cfbcb788c80a0384361b4de20420"
+COMPATIBLE_MACHINE = "chip"
```
Comment (agherzan) [inline recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb]: Usually the COMPATIBLE_MACHINE is at the end of the recipe: https://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta-skeleton/recipes-kernel/linux/linux-yocto-custom.bb .

File: recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+LED_I2C_ADDR = "0x93"
+UBOOT_FLASH_ADDR = "0x800000"
+UBI_MEMIMG_ADDR = "0x4b000000"
+UBI_FLASH_ADDR = "0x1000000"
+PAGE_SIZE = "16384"
+OOB_SIZE = "1664"
+SCRIPTADDR = "0x43100000"
+UBI_SIZE="0x0A000000"
+
+do_compile[depends] += "u-boot-chip:do_deploy"
```
Comment (agherzan) [inline recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb:24]: Wouldn't DEPEND on u-boot-chip suffice?

File: recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+NAND_FLASH_START_ADDR = "0x00000000"
+SPL_MEMIMG_ADDR = "0x44000000"
+UBOOT_MEMIMG_ADDR = "0x4a000000"
+SPL_FLASH_UPPER_ADDR = "0x400000"
+LED_I2C_CHIP = "0x34"
+LED_I2C_ADDR = "0x93"
+UBOOT_FLASH_ADDR = "0x800000"
+UBI_MEMIMG_ADDR = "0x4b000000"
+UBI_FLASH_ADDR = "0x1000000"
+PAGE_SIZE = "16384"
```
Comment (agherzan) [inline recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb]: We have this value defined in the machine configuration already. Could we use it from there (define it there and use it in both `MKUBIFS_ARGS` and here)?.

**Review by agherzan — CHANGES_REQUESTED** [review]

File: recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb (hunk 7 above)
Comment (koenkooi) [reply recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb:24]: No, DEPENDS is shorthand for "recipe:do_populate_sysroot", not "recipe:do_build"

**Comment by koenkooi** [comment]
This script is great for flashing from scratch, but it does have some drawbacks. What would be nice, but is something for a follow up PR, is the following:

1) Allow to only flash the rootfs (as in: ubi volume)
2) Use the builtin ubi tools in uboot to keep the erase counters and bad block map.

The NAND used on the CHIP is very prone to errors or just hates UBIFS, so keeping at least the bad block map would be very helpful.

**Comment by koenkooi** [comment]
And find a way to fix #11

File: recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb (hunk 7 above)
Comment (agherzan) [reply recipes-bsp/chip-u-boot-scr/chip-u-boot-scr.bb:24]: Good point.

**Comment by drewmoseley** [comment]
I pushed updated commits to the existing branch to address the review comments.  I think I made all the requested changes.  Please let me know if I missed anything.

This has also been sent to the mailing list as V3.  I guess embedding the version in the git branch and PR title is not a good approach.

**Review by agherzan — APPROVED** [review]

**Comment by agherzan** [comment]
@drewmoseley Don't worry about the branch name - it's good though to avoid it in the future. Looks good now the PR so I'll merge it. Thanks for your work.
